File: src/structs/portfolio.rs
use chrono::{DateTime, NaiveDate, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Serialize)]
pub struct PortfolioSummaryEntry {
    pub id: Uuid,
    pub member_id: i64,
    pub stock_code: String,
    pub stock_name: Option<String>,
    pub buy_date: NaiveDate,
    pub cost_per_share: f64,
    pub shares: i64,
    pub current_price: Option<f64>,
    pub current_value: Option<f64>,
    pub pnl: Option<f64>,
    pub pnl_pct: Option<f64>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Clone, Serialize, FromRow)]
pub struct PortfolioEntry {
    pub id: Uuid,
    pub member_id: i64,
    pub stock_code: String,
    pub buy_date: NaiveDate,
    pub cost_per_share: f64,
    pub shares: i64,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Deserialize)]
pub struct PortfolioRequest {
    pub stock_code: String,
    pub buy_date: NaiveDate,
    pub cost_per_share: f64,
    pub shares: i64,
}

#[derive(Serialize)]
pub struct HistoryRecord {
    pub date: NaiveDate,
    pub close: f64,
    pub adjusted_cost: f64,
    pub pnl: f64,
    pub pnl_pct: f64,
}

File: frontend/types/portfolio.ts
// Portfolio
export interface HistoryRecord {
  date: string;
  close: number;
  adjusted_cost: number;
  pnl: number;
  pnl_pct: number;
}

export interface PortfolioEntry {
  id: string;
  member_id: number;
  stock_code: string;
  buy_date: string;
  cost_per_share: number;
  shares: number;
  created_at: string;
  updated_at: string;
}

export interface PortfolioEntryInput {
  stock_code: string;
  buy_date: string;
  cost_per_share: number;
  shares: number;
}

export interface PortfolioSummaryEntry extends PortfolioEntry {
  stock_name: string;
  current_price: number | null;
  current_value: number | null;
  pnl: number | null;
  pnl_pct: number | null;
  changes: PeriodChanges;
}

export type PeriodKey = 'day' | 'week' | 'month';

/** 一筆持股在某期間的增減。基準價後端已還原期間內的除權息。 */
export interface PeriodChange {
  /** 實際比較的交易日（休市時會早於目標日） */
  base_date: string;
  base_close: number;
  /** 每股增減 */
  change: number;
  change_pct: number;
  /** 這筆持股的市值增減 */
  value_change: number;
}

/** null = 該期間沒有可用的基準日（新持股，或行情還沒補齊） */
export type PeriodChanges = Record<PeriodKey, PeriodChange | null>;
